## Topic 1. Exploratory data analysis with Pandas

### Adapted from [mlcourse.ai](https://mlcourse.ai) – Open Machine Learning Course 

Author: Yury Kashnitsky. Translated and edited by Christina Butsko, Yuanyuan Pao, Anastasia Manokhina, Sergey Isaev and Artem Trunov. This material is subject to the terms and conditions of the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license. Free use is permitted for any non-commercial purpose.


<img src="./pandas.jpg"  width=50% />

## 1. Demonstration of main Pandas methods

**[Pandas](http://pandas.pydata.org)** is a Python library that provides extensive means for data analysis. Data scientists often work with data stored in table formats like `.csv`, `.tsv`, or `.xlsx`. Pandas makes it very convenient to load, process, and analyze such tabular data using SQL-like queries. In conjunction with `Matplotlib` and `Seaborn`, `Pandas` provides a wide range of opportunities for visual analysis of tabular data.

The main data structures in `Pandas` are implemented with **Series** and **DataFrame** classes. 
- **Series**: is a one-dimensional indexed array of some fixed data type. 
- **DataFrames**: is a two-dimensional data structure - a table - where each column contains data of the same type. You can see it as a dictionary of `Series` instances. `DataFrames` are great for representing real data: rows correspond to instances (examples, observations, etc.), and columns correspond to features of these instances.

#### Loading Libraries 

First we need to load the required Python libraries. Libraries are like extensions to the base python that add functionality or help to make tasks more convenient to do. We will load some libraries that will boost your data handling capacity.

In [1]:
import numpy as np
import pandas as pd

pd.set_option("display.precision", 2)
print("libraries all imported, ready to go")

libraries all imported, ready to go


#### Loading the data 

We’ll demonstrate the main methods in action by analyzing a [dataset](https://www.kaggle.com/code/kashnitsky/topic-1-exploratory-data-analysis-with-pandas/data?select=telecom_churn.csv) on the churn rate of telecom operator clients. 

In a first step we load the dataset with pandas. To achieve this you will use the `read_csv()` method. We just need to point to the location of the dataset and indicate under what name we want to store the data,  and pandas will do the rest. 

For consistency, use the name **df** to store the output of `read_csv()` . At a first stage, the data has only been loaded. Let's have a look at the top few lines - we can use the `.head()` method to achieve this.

In [2]:
# Read the data using the read_csv function from pandas (pd) into a DataFrame named df 
# Use the format pd.read_csv - you will have to look online for the arguments
# Print out the first 5 rows of the DF using the function head()
# Try printing more rows of the DF (e.g. 10 or 20) using the function head()

df = pd.read_csv("./telecom_churn.csv")
df.head(10)

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                NaN             Yes   
1    OH             107        415                NaN             Yes   
2    NJ             137        415                NaN              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   
5    AL             118        510                Yes              No   
6    MA             121        510                 No             Yes   
7    MO             147        415                Yes              No   
8    LA             117        408                 No              No   
9    WV             141        415                Yes             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   
5                      0              223.4               98   
6                     24              218.2               88   
7                      0              157.0               79   
8                      0              184.5               97   
9                     37              258.6               84   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   
5             37.98              220.6              101             18.75   
6             37.09              348.5              108             29.62   
7             26.69              103.1               94              8.76   
8             31.37              351.6               80             29.89   
9             43.96              222.0              111             18.87   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   
7                211.8                 96                9.53   
8                215.8                 90                9.71   
9                326.4                 97               14.69   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   
7                 7.1                 6               1.92   
8                 8.7                 4               2.35   
9                11.2                 5               3.02   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  
5                       0  False  
6

Note: The last data column, **Churn**, is our target variable. It is binary: *True* indicates that that the company eventually lost this customer, and *False* indicates that the customer was retained. Later, we will build models that predict this feature based on the remaining features. This is why we call it a **target**.


In Jupyter notebooks, Pandas DataFrames are printed as these pretty tables seen above while `print(df.head())` is less nicely formatted.
By default, Pandas displays 20 columns and 60 rows, so, if your DataFrame is bigger, use the `set_option` function as shown in the example below:

```python
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 100)
```

Let’s have a look at data dimensionality using `shape` (does not take any parentheses!):

In [3]:
# Find the dimensionality of the DataFrame 
# Pass the df shape to the function print()
print(df.shape)

(3333, 20)


From the output, we can see that the table contains 3333 rows and 20 columns.

Now let’s try printing out the column names using `columns`:

In [4]:
# Print the column names of the DataFrame df using print()
print(df.columns)

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')


We can use the `info()` method to output some general information about the dataframe: 

In [5]:
# Use the info() function on the DataFrame df; pass it as an argument in print()
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3330 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

#### Missing values   
We can see from the results above that there are some missing values in our DataFrame in the feature "International plan". In `pandas`, we can also check for NAs (NULL values) using the following example: 

In [6]:
# Check for NaN under an entire DataFrame
df.isnull().values.any()

True

In [7]:
# Count the NaN under an entire DataFrame:
nan_in_df = df.isnull().sum()
nan_in_df

State                     0
Account length            0
Area code                 0
International plan        3
Voice mail plan           0
Number vmail messages     0
Total day minutes         0
Total day calls           0
Total day charge          0
Total eve minutes         0
Total eve calls           0
Total eve charge          0
Total night minutes       0
Total night calls         0
Total night charge        0
Total intl minutes        0
Total intl calls          0
Total intl charge         0
Customer service calls    0
Churn                     0
dtype: int64

In [8]:
# printing the number of values present in the whole dataframe
print('Number of NaN values present: ' + str(nan_in_df.sum()))

Number of NaN values present: 3


In the case of missing values, we should first take care to replace them in the dataset as they prevent machine learning algorithms to run. There are **many strategies** to help with missing data and they depend on whether the missing data is numeric or categorical.  What strategy is best for you problem very much depends on the specifics of your dataset. However generally speaking it is not worth to remove large chunks of data.

- simply removing rows where there is missing data (e.g. `dropna()` can achieve this)
- imputing the values with a summary statistic such as mean or median or most frequent value (e.g. `Imputer` from `sklearn` module)
- replace the values with a resonable estimate

In [9]:
# Find the instances where the time_between_orders is empty
nan_cases = df[pd.isnull(df["International plan"])]

# How many nan cases do we have? 
print(nan_cases.shape)

(3, 20)


In [10]:
# Use the unique() function on df["International plan"] and see the results

df["International plan"].unique()

array([nan, 'Yes', 'No'], dtype=object)

In [11]:
# Use the fillna() function to fill in the empty values (NAs) of the column 'International plan' 
# with a value you prefer from the unique() output above. Do not forget to assign it back (save it)
# as the new value of df['International plan']. Alternatively, you can use the argument inplace=True

df['International plan'] = df['International plan'].fillna('Yes')
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                Yes             Yes   
1    OH             107        415                Yes             Yes   
2    NJ             137        415                Yes              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

As we have seen, Pandas treats None and NaN as essentially interchangeable for indicating missing or null values. To facilitate this convention, there are several useful methods for detecting, removing, and replacing null values in Pandas data structures. They are:

- `isnull()`: Generate a boolean mask indicating missing values
- `notnull()`: Opposite of `isnull()`
- `dropna()`: Return a filtered version of the data
- `fillna()`: Return a copy of the data with missing values filled or imputed

#### Data Types

`bool`, `int64`, `float64` and `object` are the data types of our features. We see that one feature is logical (`bool`), 3 features are of type `object`, and 16 features are numeric. With this same method, we can easily see if there are any missing values. Here, there are none because each column contains 3333 observations, the same number of rows we saw before with `shape`.

We can **change the column type** with the `astype` method. Let’s apply this method to the `Churn` feature to convert it into `int64`:


In [12]:
df["Churn"] = df["Churn"].astype("int64")
df["Churn"].head()

0    0
1    0
2    0
3    0
4    0
Name: Churn, dtype: int64


The `describe` method shows basic statistical characteristics of each numerical feature (`int64` and `float64` types): number of non-missing values, mean, standard deviation, range, median, 0.25 and 0.75 quartiles.

In [13]:
# Apply the describe() function on the DataFrame
df.describe()

Account length  Area code  Number vmail messages  Total day minutes  \
count         3333.00    3333.00                3333.00            3333.00   
mean           101.06     437.18                   8.10             179.78   
std             39.82      42.37                  13.69              54.47   
min              1.00     408.00                   0.00               0.00   
25%             74.00     408.00                   0.00             143.70   
50%            101.00     415.00                   0.00             179.40   
75%            127.00     510.00                  20.00             216.40   
max            243.00     510.00                  51.00             350.80   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count          3333.00           3333.00            3333.00          3333.00   
mean            100.44             30.56             200.98           100.11   
std              20.07              9.26              50.71            19.92   
min               0.00              0.00               0.00             0.00   
25%              87.00             24.43             166.60            87.00   
50%             101.00             30.50             201.40           100.00   
75%             114.00             36.79             235.30           114.00   
max             165.00             59.64             363.70           170.00   

       Total eve charge  Total night minutes  Total night calls  \
count           3333.00              3333.00            3333.00   
mean              17.08               200.87             100.11   
std                4.31                50.57              19.57   
min                0.00                23.20              33.00   
25%               14.16               167.00              87.00   
50%               17.12               201.20             100.00   
75%               20.00               235.30             113.00   
max               30.91               395.00             175.00   

       Total night charge  Total intl minutes  Total intl calls  \
count             3333.00             3333.00           3333.00   
mean                 9.04               10.24              4.48   
std                  2.28                2.79              2.46   
min                  1.04                0.00              0.00   
25%                  7.52                8.50              3.00   
50%                  9.05               10.30              4.00   
75%                 10.59               12.10              6.00   
max                 17.77               20.00             20.00   

       Total intl charge  Customer service calls    Churn  
count            3333.00                 3333.00  3333.00  
mean                2.76                    1.56     0.14  
std                 0.75                    1.32     0.35  
min                 0.00                    0.00     0.00  
25%                 2.30                    1.00     0.00  
50%                 2.78                    1.00     0.00  
75%                 3.27                    2.00     0.00  
max                 5.40                    9.00     1.00

In order to see statistics on non-numerical features, one has to explicitly indicate data types of interest in the `include` parameter.

In [14]:
df.describe(include=["object", "bool"])

State International plan Voice mail plan
count   3333               3333            3333
unique    51                  2               2
top       WV                 No              No
freq     106               3007            2411

For categorical (type `object`) and boolean (type `bool`) features we can use the `value_counts` method. Let's take a look at the distribution of `Churn`:

In [15]:
# Use the value_counts() function just on the "Churn" column 
# Extra: if you want, you can also set the function parameter normalize to True 
# Check the function documentation online for the function arguments 

df["Churn"].value_counts()

0    2850
1     483
Name: Churn, dtype: int64

2850 users out of 3333 are *loyal*; their `Churn` value is 0. To calculate fractions, pass `normalize=True` to the `value_counts` function.


### Sorting

A `DataFrame` can be sorted by the value of one of the variables (i.e columns). For example, we can sort by *Total day charge* (use `ascending=False` to sort in descending order):


In [16]:
# Apply the function sort_values() to DataFrame df 
# Use as arguments by="Total day charge" and ascending=False 
# If you don't want to print out all of the results, follow sort_values() with head() 

df.sort_values(by="Total day charge", ascending=False).head()

State  Account length  Area code International plan Voice mail plan  \
365     CO             154        415                 No              No   
985     NY              64        415                Yes              No   
2594    OH             115        510                Yes              No   
156     OH              83        415                 No              No   
605     MO             112        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
365                       0              350.8               75   
985                       0              346.8               55   
2594                      0              345.3               81   
156                       0              337.4              120   
605                       0              335.5               77   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
365              59.64              216.5               94             18.40   
985              58.96              249.5               79             21.21   
2594             58.70              203.4              106             17.29   
156              57.36              227.4              116             19.33   
605              57.04              212.5              109             18.06   

      Total night minutes  Total night calls  Total night charge  \
365                 253.9                100               11.43   
985                 275.4                102               12.39   
2594                217.5                107                9.79   
156                 153.9                114                6.93   
605                 265.0                132               11.93   

      Total intl minutes  Total intl calls  Total intl charge  \
365                 10.1                 9               2.73   
985                 13.3                 9               3.59   
2594                11.8                 8               3.19   
156                 15.8                 7               4.27   
605                 12.7                 8               3.43   

      Customer service calls  Churn  
365                        1      1  
985                        1      1  
2594                       1      1  
156                        0      1  
605                        2      1

We can also sort by multiple columns:

In [17]:
# Can you try any examples? 
# Hint: pass a list of feature names to the argument "by" and a list of boolean values to the argument "ascending"
df.sort_values(by=["Churn", "Total day charge"], 
               ascending=[True, False]).head()

State  Account length  Area code International plan Voice mail plan  \
688     MN              13        510                 No             Yes   
2259    NC             210        415                 No             Yes   
534     LA              67        510                 No              No   
575     SD             114        415                 No             Yes   
2858    AL             141        510                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
688                      21              315.6              105   
2259                     31              313.8               87   
534                       0              310.4               97   
575                      36              309.9               90   
2858                     28              308.0              123   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
688              53.65              208.9               71             17.76   
2259             53.35              147.7              103             12.55   
534              52.77               66.5              123              5.65   
575              52.68              200.3               89             17.03   
2858             52.36              247.8              128             21.06   

      Total night minutes  Total night calls  Total night charge  \
688                 260.1                123               11.70   
2259                192.7                 97                8.67   
534                 246.5                 99               11.09   
575                 183.5                105                8.26   
2858                152.9                103                6.88   

      Total intl minutes  Total intl calls  Total intl charge  \
688                 12.1                 3               3.27   
2259                10.1                 7               2.73   
534                  9.2                10               2.48   
575                 14.2                 2               3.83   
2858                 7.4                 3               2.00   

      Customer service calls  Churn  
688                        3      0  
2259                       3      0  
534                        4      0  
575                        1      0  
2858                       1      0


### Indexing and retrieving data

A `DataFrame` can be indexed in a few different ways:  

- To get a single column, you can use a `DataFrame['Name']` construction. Let's use this to answer a question about that column alone: **what is the mean value of churned users in our dataframe?**

In [18]:
# Apply the mean() function on the column "Churn"
df["Churn"].mean()

0.14491449144914492

- **Boolean indexing** with one column is also very convenient. The syntax is `df[P(df['Name'])]`, where `P` is some logical condition that is checked for each element of the `Name` column. The result of such indexing is the `DataFrame` consisting only of rows that satisfy the `P` condition on the `Name` column. 

Let's use it to answer the question:

**How much time (on average) do churned users spend on the phone during daytime?**

In [19]:
df[df["Churn"] == 1]["Total day minutes"].mean()

206.91407867494814


**What is the maximum length of international calls among loyal users (`Churn == 0`) who do not have an international plan?**



In [20]:
# Similarly fill this in
# Hint: use a logical AND (&)

df[(df["Churn"] == 0) & (df["International plan"] == "No")]["Total intl minutes"].max()

18.9


DataFrames can be indexed by column name (label) or row name (index) or by the serial number of a row. The `loc` method is used for **indexing by name**, while `iloc()` is used for **indexing by number**.



In [21]:
# Get the values of the rows with index from 0 to 5 (inclusive) 
# and columns labeled "State" to "Area code" (inclusive). 
# Hint: use .loc

df.loc[0:5, "State":"Area code"]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

In [22]:
# Give us the values of the first five rows in the first three columns 
# (as in a typical Python slice: the maximal value is not included).
# Hint: use .iloc

df.iloc[0:5, 0:3]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

If we need the first or the last line of the data frame, we can use the `df[:1]` or `df[-1:]` construction:

In [23]:
# Try the above
df[-1:]

State  Account length  Area code International plan Voice mail plan  \
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3332             39.85              265.9               82              22.6   

      Total night minutes  Total night calls  Total night charge  \
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3332                13.7                 4                3.7   

      Customer service calls  Churn  
3332                       0      0


### Applying Functions to Cells, Columns and Rows

#### Apply function

To apply a function along an axis (rows or columns) of the DataFrame, use `apply()`. Objects passed to the function are `Series` objects whose index is either the DataFrame’s index (`axis=0`) or the DataFrame’s columns (`axis=1`). 

<img src="./apply.png"  width=50% />

In [24]:
# Use the function apply() to the DataFrame you have created
# As an argument to apply() pass the function np.max 

df.apply(np.max)

State                        WY
Account length              243
Area code                   510
International plan          Yes
Voice mail plan             Yes
Number vmail messages        51
Total day minutes         350.8
Total day calls             165
Total day charge          59.64
Total eve minutes         363.7
Total eve calls             170
Total eve charge          30.91
Total night minutes       395.0
Total night calls           175
Total night charge        17.77
Total intl minutes         20.0
Total intl calls             20
Total intl charge           5.4
Customer service calls        9
Churn                         1
dtype: object

### Extras

The `apply` method can also be used to apply a function to each row. To do this, specify `axis=1`. Lambda functions are very convenient in such scenarios. The power of lambda is better shown when you use them as an anonymous function inside another function.  For example, if we need to select all states starting with 'W', we can do it like this:

In [25]:
df[df["State"].apply(lambda state: state[0] == "W")].head()

State  Account length  Area code International plan Voice mail plan  \
9     WV             141        415                Yes             Yes   
26    WY              57        408                 No             Yes   
44    WI              64        510                 No              No   
49    WY              97        415                 No             Yes   
54    WY              87        415                 No              No   

    Number vmail messages  Total day minutes  Total day calls  \
9                      37              258.6               84   
26                     39              213.0              115   
44                      0              154.0               67   
49                     24              133.2              135   
54                      0              151.0               83   

    Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
9              43.96              222.0              111             18.87   
26             36.21              191.1              112             16.24   
44             26.18              225.8              118             19.19   
49             22.64              217.2               58             18.46   
54             25.67              219.7              116             18.67   

    Total night minutes  Total night calls  Total night charge  \
9                 326.4                 97               14.69   
26                182.7                115                8.22   
44                265.3                 86               11.94   
49                 70.6                 79                3.18   
54                203.9                127                9.18   

    Total intl minutes  Total intl calls  Total intl charge  \
9                 11.2                 5               3.02   
26                 9.5                 3               2.57   
44                 3.5                 3               0.95   
49                11.0                 3               2.97   
54                 9.7                 3               2.62   

    Customer service calls  Churn  
9                        0      0  
26                       0      0  
44                       1      0  
49                       1      0  
54                       5      1

The `map` method can be used to **replace values in a column** by passing a dictionary of the form `{old_value: new_value}` as its argument:

In [26]:
d = {"No": False, "Yes": True}

df["International plan"] = df["International plan"].map(d)
df.head()

State  Account length  Area code  International plan Voice mail plan  \
0    KS             128        415                True             Yes   
1    OH             107        415                True             Yes   
2    NJ             137        415                True              No   
3    OH              84        408                True              No   
4    OK              75        415                True              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

### GroupBy: Split, Apply, Combine

Simple aggregations can give you a flavor of your dataset, but often we would prefer to aggregate conditionally on some label or index: this is implemented in the so-called groupby operation. The name "group by" comes from a command in the SQL database language, but it is perhaps more illuminative to think of it in the terms first coined by Hadley Wickham of Rstats fame: split, apply, combine.

A canonical example of this split-apply-combine operation, where the "apply" is a summation aggregation, is illustrated in this figure:

<img src="./split-apply-combine.png"  width=50% />

This makes clear what the groupby accomplishes:

- The *split* step involves breaking up and grouping a DataFrame depending on the value of the specified key.
- The *apply* step involves computing some function, usually an aggregate, transformation, or filtering, within the individual groups.
- The *combine* step merges the results of these operations into an output array.

While this could certainly be done manually using some combination of the masking, aggregation, and merging commands, an important realization is that the intermediate splits do not need to be explicitly instantiated. Rather, the GroupBy can (often) do this in a single pass over the data, updating the sum, mean, count, min, or other aggregate for each group along the way. **The power of the GroupBy is that it abstracts away these steps: the user need not think about how the computation is done under the hood, but rather thinks about the operation as a whole.**

In general, grouping data in Pandas works as follows:



```python
df.groupby(by=grouping_columns)[columns_to_show].function()
```

1. First, the `groupby` method divides the `grouping_columns` by their values. They become a new index in the resulting dataframe.
2. Then, columns of interest are selected (`columns_to_show`). If `columns_to_show` is not included, all non groupby clauses will be included.
3. Finally, one or several functions are applied to the obtained groups per selected columns.


Apply the `groupby` method to column "Churn" from your DataFrame followed by the `apply` with an argument `np.mean`

In [27]:
# Apply the `groupby` method to column "Churn" from your DataFrame 
# followed by the `apply` with an argument `np.mean`
df.groupby(["Churn"])["Total day minutes"].apply(np.mean)

Churn
0    175.18
1    206.91
Name: Total day minutes, dtype: float64

Here is an example where we group the data according to the values of the `Churn` variable and display statistics of three columns in each group:

In [28]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].describe(percentiles=[])

Total day minutes                                   Total eve minutes  \
                  count    mean    std  min    50%    max             count   
Churn                                                                         
0                2850.0  175.18  50.18  0.0  177.2  315.6            2850.0   
1                 483.0  206.91  69.00  0.0  217.6  350.8             483.0   

                                         Total night minutes                 \
         mean    std   min    50%    max               count    mean    std   
Churn                                                                         
0      199.04  50.29   0.0  199.6  361.8              2850.0  200.13  51.11   
1      212.41  51.73  70.9  211.3  363.7               483.0  205.23  47.13   

                            
        min     50%    max  
Churn                       
0      23.2  200.25  395.0  
1      47.4  204.80  354.9

Let’s do the same thing, but slightly differently by passing a list of functions to `agg()`:

In [29]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].agg([np.mean, np.std, np.min, np.max])

Total day minutes                    Total eve minutes               \
                   mean    std amin   amax              mean    std  amin   
Churn                                                                       
0                175.18  50.18  0.0  315.6            199.04  50.29   0.0   
1                206.91  69.00  0.0  350.8            212.41  51.73  70.9   

             Total night minutes                      
        amax                mean    std  amin   amax  
Churn                                                 
0      361.8              200.13  51.11  23.2  395.0  
1      363.7              205.23  47.13  47.4  354.9


### Summary tables

Suppose we want to see how the observations in our dataset are distributed in the context of two variables - `Churn` and `International plan`. To do so, we can build a **contingency table** using the `crosstab` method:



In [30]:
# Run pd.crosstab() passing two arguments, one for each feature
# Try features Churn and "International plan" 
# Also, try the boolean argument normalize 

pd.crosstab(df["Churn"], df["International plan"])
# pd.crosstab(df["Churn"], df["Voice mail plan"], normalize=True)

International plan  False  True
Churn                          
0                    2661   189
1                     346   137

We can see that most of the users are loyal and do not use additional services (International Plan/Voice mail).

This will resemble **pivot tables** to those familiar with Excel. And, of course, pivot tables are implemented in Pandas: the `pivot_table` method takes the following parameters:

* `values` – a list of variables to calculate statistics for,
* `index` – a list of variables to group data by,
* `aggfunc` – what statistics we need to calculate for groups, ex. sum, mean, maximum, minimum or something else.

Let’s take a look at the average number of day, evening, and night calls by area code:

In [31]:
df.pivot_table(
    ["Total day calls", "Total eve calls", "Total night calls"],
    ["Area code"],
    aggfunc="mean",
)

Total day calls  Total eve calls  Total night calls
Area code                                                     
408                 100.50            99.79              99.04
415                 100.58           100.50             100.40
510                 100.10            99.67             100.60


### DataFrame transformations

Like many other things in Pandas, adding columns to a DataFrame is doable in many ways.

For example, if we want to calculate the total number of calls for all users:


In [32]:
df["Total charge"] = df["Total day charge"]+df["Total eve charge"]+df["Total night charge"]+df["Total intl charge"]
df.head()

State  Account length  Area code  International plan Voice mail plan  \
0    KS             128        415                True             Yes   
1    OH             107        415                True             Yes   
2    NJ             137        415                True              No   
3    OH              84        408                True              No   
4    OK              75        415                True              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  ...  Total eve charge  \
0             45.07              197.4  ...             16.78   
1             27.47              195.5  ...             16.62   
2             41.38              121.2  ...             10.30   
3             50.90               61.9  ...              5.26   
4             28.34              148.3  ...             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  Total charge  
0                       1      0         75.56  
1                       1      0         59.24  
2                       0      0         62.29  
3                       2      0         66.80  
4                       3      0         52.09  

[5 rows x 21 columns]

To delete columns or rows, use the `drop()` method, passing the required indexes and the `axis` parameter (`1` if you delete columns, and nothing or `0` if you delete rows). The `inplace` argument tells whether to change the original DataFrame. With `inplace=False`, the `drop` method doesn't change the existing DataFrame and returns a new one with dropped rows or columns. With `inplace=True`, it alters the DataFrame.

In [33]:
# get rid of just created columns
df.drop(["Total charge"], axis=1, inplace=True)
df.head()

State  Account length  Area code  International plan Voice mail plan  \
0    KS             128        415                True             Yes   
1    OH             107        415                True             Yes   
2    NJ             137        415                True              No   
3    OH              84        408                True              No   
4    OK              75        415                True              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

In [34]:
# And here’s how you can delete rows
df.drop([1, 2]).head()

State  Account length  Area code  International plan Voice mail plan  \
0    KS             128        415                True             Yes   
3    OH              84        408                True              No   
4    OK              75        415                True              No   
5    AL             118        510                True              No   
6    MA             121        510               False             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
3                      0              299.4               71   
4                      0              166.7              113   
5                      0              223.4               98   
6                     24              218.2               88   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   
5             37.98              220.6              101             18.75   
6             37.09              348.5              108             29.62   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   

   Customer service calls  Churn  
0                       1      0  
3                       2      0  
4                       3      0  
5                       0      0  
6                       3      0

### Mapping Categorical Features

Most ML libraries and tools will only accept *numerical values* as their input. In the case where we have categorical features present, we need to represent them as numerical values. Unlike the class label, when dealing with categorical input features, typically one converts each categorical feature using “one-hot encoding”. The input in one-hot encoding is the vector of discrete categorical values, and the output will be a sparse matrix where each column corresponds to one possible value of one feature.

In our example, the feature State is a categorical feature with values such as ["KS", "OH", "NJ", "OK"] etc. Such features can be easily mapped to dummy variables which could be expressed as [0, 1, 2]. Can you spot any problem though with this approach? Even though the country values do not come in any particular order, a machine learning algorithm will now assume that "KS" is larger than "OH", and "OH" is larger than "NJ", and so on. Although this assumption is incorrect, the algorithm could still produce useful results. However, those results would not be optimal.

The correct approach in this case is to apply one-hot encoding. This estimator transforms each unique categorical value of a single input categorical feature to a new dummy feature. So, for our 51 unique country values , we will end up with 51 new dummy features after one-hot-encoding. There are plenty of libraries and functions that are used for one-hot encoding. In this example, we will use the `get_dummies()` function from pandas:

In [35]:
# Use the function unique() on the df['State'] column 

df['State'].unique()

array(['KS', 'OH', 'NJ', 'OK', 'AL', 'MA', 'MO', 'LA', 'WV', 'IN', 'RI',
       'IA', 'MT', 'NY', 'ID', 'VT', 'VA', 'TX', 'FL', 'CO', 'AZ', 'SC',
       'NE', 'WY', 'HI', 'IL', 'NH', 'GA', 'AK', 'MD', 'AR', 'WI', 'OR',
       'MI', 'DE', 'UT', 'CA', 'MN', 'SD', 'NC', 'WA', 'NM', 'NV', 'DC',
       'KY', 'ME', 'MS', 'TN', 'PA', 'CT', 'ND'], dtype=object)

In [36]:
# Use the function nunique() on the df['State'] column 

df['State'].nunique()

51

In [37]:
# Apply one-hot encoding to the categorical feature Country
pd.set_option('display.max_columns', 100)

states = df['State']
states = pd.get_dummies(states)

print(states.shape)

states.head(10)

(3333, 51)


AK  AL  AR  AZ  CA  CO  CT  DC  DE  FL  GA  HI  IA  ID  IL  IN  KS  KY  LA  \
0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   
1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
4   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
5   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
6   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
7   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
8   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   
9   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

   MA  MD  ME  MI  MN  MO  MS  MT  NC  ND  NE  NH  NJ  NM  NV  NY  OH  OK  OR  \
0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   
2   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   
3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   
4   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   
5   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
6   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
7   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   
8   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
9   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

   PA  RI  SC  SD  TN  TX  UT  VA  VT  WA  WI  WV  WY  
0   0   0   0   0   0   0   0   0   0   0   0   0   0  
1   0   0   0   0   0   0   0   0   0   0   0   0   0  
2   0   0   0   0   0   0   0   0   0   0   0   0   0  
3   0   0   0   0   0   0   0   0   0   0   0   0   0  
4   0   0   0   0   0   0   0   0   0   0   0   0   0  
5   0   0   0   0   0   0   0   0   0   0   0   0   0  
6   0   0   0   0   0   0   0   0   0   0   0   0   0  
7   0   0   0   0   0   0   0   0   0   0   0   0   0  
8   0   0   0   0   0   0   0   0   0   0   0   0   0  
9   0   0   0   0   0   0   0   0   0   0   0   1   0

At this stage, we want to update in our Dataframe the categorical feature "State" with the result of one-hot encoding. Remember, that are going to remove a single feature, and add 51 new dummy variables created by one-hot encoding.

In [38]:
# Remove and replace the Country values from customers with the result of one-hot encoding

df = df.drop('State', axis = 1)
df = df.join(states)
df.head()

Account length  Area code  International plan Voice mail plan  \
0             128        415                True             Yes   
1             107        415                True             Yes   
2             137        415                True              No   
3              84        408                True              No   
4              75        415                True              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  AK  AL  AR  AZ  CA  CO  CT  DC  DE  FL  GA  \
0                       1      0   0   0   0   0   0   0   0   0   0   0   0   
1                       1      0   0   0   0   0   0   0   0   0   0   0   0   
2                       0      0   0   0   0   0   0   0   0   0   0   0   0   
3                       2      0   0   0   0   0   0   0   0   0   0   0   0   
4                       3      0   0   0   0   0   0   0   0   0   0   0   0   

   HI  IA  ID  IL  IN  KS  KY  LA  MA  MD  ME  MI  MN  MO  MS  MT  NC  ND  NE  \
0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   
1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
4   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

   NH  NJ  NM  NV  NY  OH  OK  OR  PA  RI  SC  SD  TN  TX  UT  VA  VT  WA  WI  \
0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
1   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   
2   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
3   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   
4   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   

   WV  WY  
0   0   0  
1   0   0  
2   0   0  
3   0   0  
4   0   0

We can convert the values "Yes"/"No" of the binary features into numerical values using [`pandas.Series.map()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.map.html):

In [39]:
df["International plan"] = df["International plan"].map({"Yes": 1, "No": 0})
df["Voice mail plan"]    = df["Voice mail plan"].map({"Yes": 1, "No": 0})
df.head()

Account length  Area code  International plan  Voice mail plan  \
0             128        415                 NaN                1   
1             107        415                 NaN                1   
2             137        415                 NaN                0   
3              84        408                 NaN                0   
4              75        415                 NaN                0   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  AK  AL  AR  AZ  CA  CO  CT  DC  DE  FL  GA  \
0                       1      0   0   0   0   0   0   0   0   0   0   0   0   
1                       1      0   0   0   0   0   0   0   0   0   0   0   0   
2                       0      0   0   0   0   0   0   0   0   0   0   0   0   
3                       2      0   0   0   0   0   0   0   0   0   0   0   0   
4                       3      0   0   0   0   0   0   0   0   0   0   0   0   

   HI  IA  ID  IL  IN  KS  KY  LA  MA  MD  ME  MI  MN  MO  MS  MT  NC  ND  NE  \
0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   
1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
4   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

   NH  NJ  NM  NV  NY  OH  OK  OR  PA  RI  SC  SD  TN  TX  UT  VA  VT  WA  WI  \
0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
1   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   
2   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
3   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   
4   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   

   WV  WY  
0   0   0  
1   0   0  
2   0   0  
3   0   0  
4   0   0

## Useful resources

* The same notebook as an interactive web-based [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-exploratory-data-analysis-with-pandas)
* ["Merging DataFrames with pandas"](https://nbviewer.jupyter.org/github/Yorko/mlcourse.ai/blob/master/jupyter_english/tutorials/merging_dataframes_tutorial_max_palko.ipynb) - a tutorial by Max Plako within mlcourse.ai (full list of tutorials is [here](https://mlcourse.ai/tutorials))
* ["Handle different dataset with dask and trying a little dask ML"](https://nbviewer.jupyter.org/github/Yorko/mlcourse.ai/blob/master/jupyter_english/tutorials/dask_objects_and_little_dask_ml_tutorial_iknyazeva.ipynb) - a tutorial by Irina Knyazeva within mlcourse.ai
* Main course [site](https://mlcourse.ai), [course repo](https://github.com/Yorko/mlcourse.ai), and YouTube [channel](https://www.youtube.com/watch?v=QKTuw4PNOsU&list=PLVlY_7IJCMJeRfZ68eVfEcu-UcN9BbwiX)
* Official Pandas [documentation](http://pandas.pydata.org/pandas-docs/stable/index.html)
* Course materials as a [Kaggle Dataset](https://www.kaggle.com/kashnitsky/mlcourse)
* Medium ["story"](https://medium.com/open-machine-learning-course/open-machine-learning-course-topic-1-exploratory-data-analysis-with-pandas-de57880f1a68) based on this notebook
* If you read Russian: an [article](https://habrahabr.ru/company/ods/blog/322626/) on Habr.com with ~ the same material. And a [lecture](https://youtu.be/dEFxoyJhm3Y) on YouTube
* [10 minutes to pandas](http://pandas.pydata.org/pandas-docs/stable/10min.html)
* [Pandas cheatsheet PDF](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
* GitHub repos: [Pandas exercises](https://github.com/guipsamora/pandas_exercises/) and ["Effective Pandas"](https://github.com/TomAugspurger/effective-pandas)
* [scipy-lectures.org](http://www.scipy-lectures.org/index.html) — tutorials on pandas, numpy, matplotlib and scikit-learn
